source/opt/cfg_cleanup_pass.cpp

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a pass to cleanup the CFG to remove superfluous
+// constructs (e.g., unreachable basic blocks, empty control flow structures,
+// etc)
+
+#include <queue>
+#include <unordered_set>
+
+#include "cfg_cleanup_pass.h"
+
+#include "function.h"
+#include "module.h"
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+
+// Remove Phi operand {|rm_arg_id|, |rm_block_id|} from |phi|. This works by
+// building a new operand vector for |phi|.  The new operand vector will have
+// all the original operands except the operand {|rm_arg_id|, |rm_block_id|}.
+void RemovePhiOperand(ir::Instruction* phi, uint32_t rm_arg_id,
+                      uint32_t rm_block_id) {
+  std::vector<ir::Operand> keep_operands;
+
+  // Traverse all the operands in |phi|. Build the new operand vector by adding
+  // all the original operands from |phi| except the unwanted one.
+  for (uint32_t i = 0; i < phi->NumOperands();) {
+    if (i < 2) {
+      // The first two arguments are always preserved.
+      keep_operands.push_back(phi->GetOperand(i));
+      ++i;
+    } else {
+      // The remaining Phi arguments come in pairs. Index 'i' contains the
+      // variable id, index 'i + 1' is the originating block id. If these two
+      // match {|rm_arg_id|, |rm_block_id|}, skip them.
+      assert(i % 2 == 0 && i < phi->NumOperands() - 1 &&
+             "malformed Phi arguments");
+      if (rm_arg_id != phi->GetSingleWordOperand(i) ||
+          rm_block_id != phi->GetSingleWordOperand(i + 1)) {
+        keep_operands.push_back(phi->GetOperand(i));
+        keep_operands.push_back(phi->GetOperand(i + 1));
+      }
+
+      i += 2;
+    }
+  }
+
+  phi->ReplaceOperands(keep_operands);
+}
+
+}  // namespace
+
+void CFGCleanupPass::RemoveFromReachedPhiOperands(const ir::BasicBlock& block,
+                                                  ir::Instruction* inst) {
+  uint32_t inst_id = inst->result_id();
+  if (inst_id == 0) {
+    return;
+  }
+
+  analysis::UseList* uses = def_use_mgr_->GetUses(inst_id);
+  if (uses == nullptr) {
+    return;
+  }
+
+  for (auto u : *uses) {
+    if (u.inst->opcode() == SpvOpPhi) {
+      RemovePhiOperand(u.inst, inst_id, block.id());
```

Comment:
In SPIR-V the block in a Phi is the immediate predecessor block, not necessarily the block in which the associated value is defined.

Example:

%block0  defines value %e
Then %block0 branches into %block1, then %block1 branches to %block2. 
%block2 has a phi with (%e, %block1).
Then when killing %block0 you're going to delete the definition of %e but miss removing (%e, %block1) from the phi.

So you need an alternative case where you substitute an undef value for inst used in a subsequent phi.
E.g. something like in https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/opt/local_ssa_elim_pass.cpp#L271 where an undef value is generated for the given type.

You'll need a test case for this scenario.